File: ungs-docsys-bff/src/modules/job-applications/dtos/job-application-request.dto.ts
import {
    IsNotEmpty,
    IsString,
    MaxLength,
    IsNumber,
    Min,
    Max,
    IsPositive,
    IsOptional
} from 'class-validator';
import { Type } from 'class-transformer';
import { RequirementCreateRequestDto } from '../requirements/dtos/requirement-create-request.dto';
import { ApiProperty, ApiPropertyOptional } from '@nestjs/swagger';

export class JobApplicationRequestDto {
    @ApiProperty()
    @IsNotEmpty({ message: 'Title is required' })
    @IsString()
    @MaxLength(150, { message: 'Title must not exceed 150 characters' })
    title: string;
  
    @ApiProperty()
    @IsNotEmpty({ message: 'Description is required' })
    @IsString()
    @MaxLength(5000, { message: 'Description must not exceed 5000 characters' })
    description: string;
  
    @ApiProperty()
    @IsNotEmpty({ message: 'Job Application Period ID is required' })
    @IsNumber()
    @IsPositive({ message: 'Job Application Period ID must be a positive number' })
    jobApplicationPeriodId: number;
  
    @ApiProperty()
    @IsNotEmpty({ message: 'Minimum Approvers is required' })
    @IsNumber()
    @Min(1, { message: 'Minimum Approvers must be at least one approver' })
    @Max(5, { message: 'Minimum Approvers must not exceed 5' })
    minApprovers: number;
  
    @ApiProperty()
    @IsNotEmpty({ message: 'Reason is required' })
    @IsString()
    @MaxLength(500, { message: 'Reason must not exceed 500 characters' })
    reason: string;
  
    @ApiProperty()
    @IsNotEmpty({ message: 'Year Period is required' })
    @IsNumber()
    @Type(() => Number)
    yearPeriod: number;

    @ApiProperty()
    @IsNotEmpty({ message: 'Year Period is required' })
    @IsNumber()
    @Type(() => Number)
    jobProfileLevelId: number;

    @ApiProperty()
    @IsOptional()
    @IsNumber()
    @Type(() => Number)
    jobApplicationStatusId?: number;

    @ApiPropertyOptional({
        description: 'List of requirements associated with the job application',
        type: [RequirementCreateRequestDto],
    })
    @IsOptional()
    requirements?: RequirementCreateRequestDto[];
}

File: ungs-docsys-ms/src/main/java/com/ungs/docsys/dtos/JobApplicationRequestDto.java
package com.ungs.docsys.dtos;

import jakarta.validation.constraints.*;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.util.List;

@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class JobApplicationRequestDto {

    @NotBlank(message = "Title is required")
    @Size(max = 150, message = "Title must not exceed 150 characters")
    private String title;

    @NotBlank(message = "Description is required")
    @Size(max = 5000, message = "Description must not exceed 5000 characters")
    private String description;

    @NotNull(message = "Job Application Period ID is required")
    @Positive(message = "Job Application Period ID must be a positive number")
    private Long jobApplicationPeriodId;

    @NotNull(message = "Minimum Approvers is required")
    @Min(value = 1, message = "Minimum Approvers must be at least one approver")
    @Max(value = 5, message = "Minimum Approvers must not exceed 5")
    private Long minApprovers;

    @NotBlank(message = "Reason is required")
    @Size(max = 500, message = "Reason must not exceed 500 characters")
    private String reason;

    @NotNull(message = "Year Period is required")
    private Long yearPeriod;

    @NotNull(message = "Job profile leve is required")
    private Long jobProfileLevelId;

    private Long jobApplicationStatusId;

    private List<RequirementRequestDto> requirements;
}
